src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -755,14 +778,30 @@ private async Task AcquireWriteLockAsync(CancellationToken cancellationToken)
             if (!acquireLockTask.IsCompletedSuccessfully)
             {
                 Interlocked.Increment(ref _pendingWriters);
+
                 try
                 {
                     await acquireLockTask.ConfigureAwait(false);
                 }
-                finally
+                catch
                 {
-                    Interlocked.Decrement(ref _pendingWriters);
+                    if (Interlocked.Decrement(ref _pendingWriters) == 0)
+                    {
+                        // If a pending waiter is canceled, we may end up in a situation where a previously written frame
+                        // saw that there were pending writers and as such deferred its flush to them, but if/when that pending
+                        // writer is canceled, nothing may end up flushing the deferred work (at least not promptly).  To compensate,
+                        // if a pending writer does end up being canceled, we queue a flush.  We can't check whether there's such a pending
+                        // operation because we failed to acquire the lock that protects that state.  But we can at least only
+                        // do the queue if our decrement caused the pending count to reach 0: if it's still higher than zero,
+                        // then there's at least one other pending writer who can handle the flush.  Worst case, we queue a flush
+                        // that ends up being a nop.
+                        ThreadPool.UnsafeQueueUserWorkItem(thisRef => thisRef.LogExceptions(thisRef.FlushAsync()), this, preferLocal: false);
```

Comment:
Do we actually need to queue this? Seems like we could just invoke FlushAsync but not await it.

Reply:
I think you're right.  I don't like all the fire-and-forget that's made its way into the code base, but this isn't making it worse.